src/term/transform/float_combinators.rs

```diff
@@ -1,277 +1,180 @@
 
-      Detach::Recursive if arg_scope.is_empty() && arg_detach.can_detach() => {
-        term_info.detach_term(self);
-        Detach::Combinator
-      }
-
-      // If the only term that is possible to detach is just a Term::Ref,
-      // there is no benefit to it, so that case is skipped
-      Detach::Recursive if !matches!(fun, box Term::Ref { .. }) => {
-        term_info.detach_term(fun);
-        arg_detach
-      }
-
-      _ => fun_detach & arg_detach,
-    };
-
-    term_info.merge_scope(fun_scope);
-    term_info.merge_scope(arg_scope);
-
-    detach
+    declared.difference(&used).count() != 0 || used.difference(&declared).count() != 0
   }
 
-  fn float_mat(&mut self, depth: usize, term_info: &mut TermInfo) -> Detach {
-    let Term::Mat { args, rules } = self else { unreachable!() };
-
-    let mut detach = Detach::Combinator;
-    for arg in args {
-      detach = detach & arg.go_float(depth + 1, term_info);
-    }
-    let parent_scope = term_info.replace_scope(BTreeSet::new());
-
-    for rule in rules.iter_mut() {
-      for pat in &rule.pats {
-        debug_assert!(pat.is_native_num_match());
-      }
-
-      let arm_detach = match rule.body.go_float(depth + 1, term_info) {
-        // If the recursive ref reached here, it is not in a active position
-        Detach::Recursive => Detach::Combinator,
-        detach => detach,
-      };
-
-      detach = detach & arm_detach;
-    }
+  fn is_combinator(&self) -> bool {
+    self.is_closed()
+  }
 
-    term_info.merge_scope(parent_scope);
-    detach
+  pub fn is_closed(&self) -> bool {
+    self.free_vars().is_empty() && !self.has_unscoped_diff() && !matches!(self, Term::Ref { .. })
   }
+}
 
-  // We don't want to detach id function, since that's not a net gain in performance or space
-  fn is_id(&self) -> bool {
+impl Term {
+  pub fn float_children_mut(&mut self) -> impl DoubleEndedIterator<Item = &mut Term> {
+    multi_iterator!(FloatIter { Zero, One, Two, Vec, Mat, App });
     match self {
-      Term::Lam { nam: Some(lam_nam), bod: box Term::Var { nam: var_nam }, .. } => lam_nam == var_nam,
-      _ => false,
+      Term::App { fun, arg, .. } => {
+        let mut args = vec![arg.as_mut()];
+        let mut app = fun.as_mut();
+        while let Term::App { fun, arg, .. } = app {
+          args.push(arg);
+          app = fun;
+        }
+        args.push(app);
+        FloatIter::App(args)
+      }
+      Term::Mat { args, rules } => {
+        FloatIter::Mat(args.iter_mut().chain(rules.iter_mut().map(|r| &mut r.body)))
+      }
+      Term::Tup { els } | Term::Sup { els, .. } | Term::Lst { els } => FloatIter::Vec(els),
+      Term::Let { val: fst, nxt: snd, .. }
+      | Term::Use { val: fst, nxt: snd, .. }
+      | Term::Dup { val: fst, nxt: snd, .. }
+      | Term::Opx { fst, snd, .. } => FloatIter::Two([fst.as_mut(), snd.as_mut()]),
+      Term::Lam { bod, .. } | Term::Chn { bod, .. } => FloatIter::One([bod.as_mut()]),
```

Comment:
If you change the Lam case to be `bod.float_children_mut()` I think it'll always keep sequences of lambdas together which would avoid fragmenting some combinators